agc/asm.c

```diff
@@ -0,0 +1,399 @@
+/*
+  Copyright 2003-2006 Ronald S. Burkey <info@sandroid.org>,
+            2008 Onno Hommes
+            2016 Inokentiy Babushkin
+
+  This file is based on the implmentation found in yaAGC, but modified for new
+  purposes.
+
+  This file is part of yaAGC / radare2-extras.
+  yaAGC is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2 of the License, or
+  (at your option) any later version.
+  yaAGC is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+  You should have received a copy of the GNU General Public License
+  along with yaAGC; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+
+  Filename:  asm_agc.c
+  Purpose:   Source file for AGC Disassembler integrated in radare2.
+  Contact:   Inokentiy Babushkin
+  Reference: http://www.ibiblio.org/apollo
+  Mods:      08/31/08 OH. Began.
+             07/12/16 IB. Began refactoring to include in radare2-extras.
+             07/17/16 IB. Annotated disassembler source and added symbolic
+                          constants.
+*/
+
+#include <stdbool.h>
+#include "asm.h"
+
+const char *agc_mnemonics[59] = {
+    // 3-bit opcode, no operand
+    "com", "ddoubl", "double", "dtcb", "dtcf", "extend", "inhint", "noop",
+    "ovsk", "relint", "resume", "return", "tcaa", "xlq", "xxalq", "zl",
+    // 3-bit opcode, w/ operand
+    "tc %04o", "ccs %04o", "tcf %04o", "das %04o", "lxch %04o",
+    "incr %04o", "ads %04o", "ca %04o", "cs %04o", "index %04o",
+    "dxch %04o", "ts %04o", "xch %04o", "ad %04o", "mask %04o",
+    // 4-bit opcode, no operand
+    "dcom", "resume", "square", "zq",
+    // 4-bit opcode, w/ small operand
+    "read %03o", "write %03o", "rand %03o", "wand %03o", "ror %03o",
+    "wor %03o", "rxor %03o", "edrupt %03o",
+    // 4-bit opcode, w/ normal operand
+    "dv %04o", "bzf %04o", "msu %04o", "qxch %04o", "aug %04o",
+    "dim %04o", "dca %04o", "dcs %04o", "index %04o", "su %04o",
+    "bzmf %04o", "mp %04o"
+};
+
+// decode a 3-bit opcode
+static void decode_no_extra_opcode_bit(
+        agc_insn_t *insn, ut16 value, ut16 nop_value) {
+    // first, special cases that have their own mnemonics
+    if (value == 040000) { // i = 100, o = 0
```

Comment:
this if/else chain looks pretty unmaintainable. maybe a table of "value, type" should be enough